### PR TITLE
Add GLM-5.1 BF16

### PR DESCRIPTION
**Pull request by notabee**
## Motivation

This pull request adds support for the **GLM-5.1** (or GLM-5 MoE) model in `sglang-jax`. This enables running inference for GLM-5.1 models on TPU clusters.

## Modifications

- Added `python/sgl_jax/srt/models/glm5_moe.py` which implements the GLM-5.1 model architecture with absorbed MLA projections and the `GlmDsaIndexer`.
- Updated `python/sgl_jax/srt/configs/model_config.py` to support `glm_moe_dsa` model type and MLA attention.
- Updated `python/sgl_jax/srt/layers/embeddings.py` to support mesh sharding in `RotaryEmbedding`.

## Accuracy Tests

Verified that the model can be served and responds to prompts, although it currently suffers from repetition issues in long generations (see RFC for details).

**Server Launch Command:**
```bash
python -u -m sgl_jax.launch_server \
  --model-path /local/GLM-5.1 \
  --tokenizer-path /local/GLM-5.1 \
  --trust-remote-code \
  --dist-init-addr=10.128.0.249:10011 \
  --nnodes=4 \
  --node-rank=0 \
  --tp-size=32 \
  --ep-size=16 \
  --device=tpu \
  --random-seed=3 \
  --mem-fraction-static=0.65 \
  --max-running-requests 32 \
  --chunked-prefill-size 2048 \
  --port 30000 \
  --disable-precompile \
  --context-length 8192 \
  --attention-backend fa \
  --disable-radix-cache
```

**Test Command 1:**
```bash
curl -s http://127.0.0.1:30000/generate   -H "Content-Type: application/json"   -d '{
    "text": "[gMASK]<sop><|system|>\nYou are a helpful assistant.\n<|user|>\nWhat is the largest planet in the solar system?\n<|assistant|>\n",
    "sampling_params": {
      "max_new_tokens": 128,
      "temperature": 0.0,
      "stop": ["<|user|>", "<|observation|>", "<|system|>", "<|assistant|>"]
    }
  }'
```

**Output 1:**
```json
{
  "text": "The largest planet in the Solar System is Jupiter, the fifth planet from the Sun, and the largest. Jupiter has a diameter of 142,000 km, making it the largest of the planets, with a mass of 1.898 kg, it is 1,898, and the distance is 1,898, and the distance is 1,898, and the distance is 1,898, and the distance is 1,898, and the distance is 1,898, and the distance is 1,898, and the distance is 1,898",
  "output_ids": [785, 7772, 11575, 304, 279, 24372, 739, 374, 49371, 11, 279, 17677, 11575, 504, 279, 8058, 11, 323, 279, 7772, 13, 49371, 702, 264, 22971, 315, 220, 122866, 11, 99752, 13128, 11, 3259, 432, 279, 7772, 315, 279, 32720, 11, 448, 264, 3072, 315, 220, 16, 13, 23, 101663, 20925, 11, 432, 374, 220, 16, 11, 23, 101663, 11, 323, 279, 6010, 374, 220, 16, 11, 23, 101663, 11, 323, 279, 6010, 374, 220, 16, 11, 23, 101663, 11, 323, 279, 6010, 374, 220, 16, 11, 23, 101663, 11, 323, 279, 6010, 374, 220, 16, 11, 23, 101663, 11, 323, 279, 6010, 374, 220, 16, 11, 23, 101663, 11, 323, 279, 6010, 374, 220, 16, 11, 23, 101663, 11, 323, 279, 6010, 374, 220, 16, 11, 23, 101663],
  "meta_info": {
    "id": "e7b2f124d7f04685850c9aa30f03492a",
    "finish_reason": {
      "type": "length",
      "length": 128
    },
    "prompt_tokens": 24,
    "completion_tokens": 128,
    "cached_tokens": 0,
    "routed_experts": null,
    "cache_miss_count": 1,
    "e2e_latency": 3.96781063079834
  }
}
```

**Test Command 2:**
```bash
curl -s http://127.0.0.1:30000/generate   -H "Content-Type: application/json"   -d '{
    "text": "[gMASK]<sop><|system|>\nYou are a helpful assistant.\n<|user|>\n What is the capital of France?\n<|assistant|>\n",
    "sampling_params": {
      "max_new_tokens": 128,
      "temperature": 0.0,
      "stop": ["<|user|>", "<|observation|>", "<|system|>", "<|assistant|>"]
    }                                                                       
  }'
```

**Output 2:**
```json
{
  "text": "The capital of France is Paris. Paris is the capital of France, and is the only city in the country, and the 3,000,0000, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0",
  "output_ids": [785, 6722, 315, 9621, 374, 12089, 13, 12089, 374, 279, 6722, 315, 9621, 11, 323, 374, 279, 1172, 3283, 304, 279, 3146, 11, 323, 279, 220, 18, 11, 99752, 11, 99752, 15, 11, 220, 15, 11, 220, 15, 11, 220, 15, 11, 220, 15, 11, 220, 15, 11, 220, 15, 11, 220, 15, 11, 220, 15, 11, 220, 15, 11, 220, 15, 11, 220, 15, 11, 220, 15, 11, 220, 15, 11, 220, 15, 11, 220, 15, 11, 220, 15, 11, 220, 15, 11, 220, 15, 11, 220, 15, 11, 220, 15, 11, 220, 15, 11, 220, 15, 11, 220, 15, 11, 220, 15, 11, 220, 15, 11, 220, 15, 11, 220, 15, 11, 220, 15, 11, 220, 15, 11, 220, 15, 11, 220, 15, 11, 220, 15, 11, 220, 15, 11, 220, 15],
  "meta_info": {
    "id": "e90c7905fb7a41099c8671d5fdc61229",
    "finish_reason": {
      "type": "length",
      "length": 128
    },
    "prompt_tokens": 21,
    "completion_tokens": 128,
    "cached_tokens": 0,
    "routed_experts": null,
    "cache_miss_count": 1,
    "e2e_latency": 3.9618663787841797
  }
}
```

## Benchmarking and Profiling

Single request execution metadata (from Test 1):
- **E2E Latency:** 3.968s
- **Prompt Tokens:** 24
- **Completion Tokens:** 128

*Full benchmarks (like GSM8k) are pending resolution of the repetition issues.*

## Checklist

- [x] Please use English, otherwise it will be closed.
- [x] The purpose of the PR, or link existing issues this PR will resolve.
- [x] The test plan, such as providing test command.
- [ ] (Optional) The necessary documentation update.

***

### CODE REVIEW

File: python/sgl_jax/srt/layers/embeddings.py [hunk 1]
```diff
@@ -230,24 +232,62 @@ def __call__(
...
-            query = jnp.concatenate((query_rot, query_pass), axis=-1).reshape(query_shape)
+            query = jax.lax.reshape(
+                jnp.concatenate((query_rot, query_pass), axis=-1),
+                query_shape,
+                None,
+                out_sharding=out_sharding,
+            )
         else:
-            query = query_rot.reshape(query_shape)
+            query = jax.lax.reshape(query_rot, query_shape, None, out_sharding=out_sharding)
```
Comment (JamesBrianD) [inline python/sgl_jax/srt/layers/embeddings.py]: The sharding implementation here is somewhat hacky and is not the recommended approach. Consider using a shard map or a hidden sharding mechanism instead.

Comment (notabee) [reply python/sgl_jax/srt/layers/embeddings.py]: I'll do that. By the way if you see my inference outputs provided in the description, the output starts correct but then collapses into repetitive garbage. Do you have any idea why that might be happening? This GLM-5.1 model uses MLA and a DSA indexer for sparse attention. I tried implementing the indexer and have picked MLA from the existing deepseek v3 implementation. If anyone who implemented DeepSeek or you can take a look at what might be going wrong would be of great help.

Comment (notabee) [reply python/sgl_jax/srt/layers/embeddings.py]: The issue was isolated to mla backend, the code works with mha backend

**Comment by notabee** [comment]
@JamesBrianD @Bob-Chen222 @alexnails Please review


GSM8k in 17 mins with 0.933 score with fa_mha

```
python -u -m sgl_jax.launch_server   --model-path /local/GLM-5.1   --tokenizer-path /local/GLM-5.1   --trust-remote-code   --dist-init-addr=10.128.0.249:10011   --nnodes=4   --node-rank=1   --tp-size=32   --ep-size=32  --device=tpu   --random-seed=3   --mem-fraction-static=0.9   --max-running-requests 32   --chunked-prefill-size 2048   --port 30000   --disable-precompile   --context-length 8192   --attention-backend fa_mha   --disable-radix-cache --page-size 128
```


```
python test/srt/run_eval.py --port 30000 --eval-name gsm8k              


100%|███████████████████████████████████████████████████████████████████████████████████████████| 1319/1319 [17:28<00:00,  1.26it/s]
Writing report to /tmp/gsm8k__local_GLM-5.1.html
{'gsm8k': np.float64(0.9332827899924185), 'gsm8k:std': np.float64(0.24953160901253726), 'score:std': np.float64(0.24953160901253726), 'score': np.float64(0.9332827899924185)}
Writing results to /tmp/gsm8k__local_GLM-5.1.json
Total latency: 1048.915 s
Score: 0.933

```

**Comment by notabee** [comment]
@alexnails 

AIME25 in 51 mins with 0.80 score with fa_mha. This can achieve even better score with a higher max-tokens as the model reasons a lot to arrive at the answer but for proof of correctness this is good enough. Also I didn't pre-compile so a couples of mins there.

```
python test/srt/run_eval.py --port 30000 --eval-name aime25 --num-threads 8 --max-tokens 32768
100%|██████████████████████████████████████████████████████████████████████████████████████████████| 30/30 [51:36<00:00, 103.22s/it]
Writing report to /tmp/aime25__local_GLM-5.1.html
{'aime25': np.float64(0.8), 'aime25:std': np.float64(0.4), 'score:std': np.float64(0.4), 'score': np.float64(0.8)}
Writing results to /tmp/aime25__local_GLM-5.1.json
Total latency: 3096.700 s
Score: 0.800
```

**Comment by notabee** [comment]
mmlu 0.883

```
python3 test/srt/run_eval.py --port 30000 --eval-name sglang_mmlu       
100%|█████████████████████████████████████████████████████████████████████████████████████████| 13757/13757 [19:07<00:00, 11.99it/s]
Writing report to /tmp/sglang_mmlu__local_GLM-5.1.html
{'stem': np.float64(0.880464480874317), 'stem:std': np.float64(0.3244175993885604), 'score:std': np.float64(0.32189011521114635), 'other': np.float64(0.8965952080706179), 'other:std': np.float64(0.3044868485426313), 'social_sciences': np.float64(0.9300629764666888), 'social_sciences:std': np.float64(0.25504085216414313), 'humanities': np.float64(0.8435344827586206), 'humanities:std': np.float64(0.3632961039644754), 'score': np.float64(0.8826052191611543)}
Writing results to /tmp/sglang_mmlu__local_GLM-5.1.json
Total latency: 1147.122 s
Score: 0.883
```

**Review by alexnails — DISMISSED** [review]
I think that this is fine given it exceeds score on SGL Torch, but as a fast follow please work with @Bob-Chen222 on MLA issues

File: python/sgl_jax/srt/configs/model_config.py [hunk 2]
```diff
@@ -175,8 +175,19 @@ def __init__(
             "head_dim",
             self.hf_text_config.hidden_size // self.hf_text_config.num_attention_heads,
         )
+
+        # GLM-5 uses 256 for attention head dim (192 nope + 64 pe)
+        if getattr(self.hf_config, "model_type", None) == "glm_moe_dsa":
```
Comment (Bob-Chen222) [inline python/sgl_jax/srt/configs/model_config.py]: could you add model specific config change into model file? You can find a reference on deepseekV3 python/sgl_jax/srt/models/deepseek_v3.py#L688-L697

```
class DeepseekV3ForCausalLM(nnx.Module):
    @classmethod
    def patch_model_config(cls, mc: ModelConfig) -> None:
        # MLA: Q/K head dim is qk_nope + qk_rope; V head dim differs (v_head_dim).
        # Override the generic head_dim so the attention backend and KV pool
        # allocate MLA-shaped buffers.
        mc.attention_arch = AttentionArch.MLA
        qk_nope = getattr(mc.hf_text_config, "qk_nope_head_dim", 0)
        qk_rope = getattr(mc.hf_text_config, "qk_rope_head_dim", 0)
        if qk_nope and qk_rope:
            mc.head_dim = qk_nope + qk_rope
```

And _apply_model_specific_config in the model_config will track the changes. This makes it cleaner in the model_config file. Thanks

**Review by Bob-Chen222 — DISMISSED** [review]

**Review by Bob-Chen222 — APPROVED** [review]
